<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>cellGeometry: Quick Start Guide • cellGeometry</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="cellGeometry: Quick Start Guide"></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">cellGeometry</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.5.8</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/intro.html">Get Started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/algorithm.html">Mathematical Foundations</a></li>
    <li><a class="dropdown-item" href="../articles/visualization.html">Visualization Guide</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><a class="dropdown-item" href="../articles/index.html">All articles</a></li>
  </ul></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-tutorials" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Tutorials</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-tutorials"><li><a class="dropdown-item" href="../articles/intro.html">Quick Start Guide</a></li>
    <li><a class="dropdown-item" href="../articles/real_world.html">Real-World Application</a></li>
    <li><a class="dropdown-item" href="../articles/brain_atlas.html">Large-Scale: Brain Atlas</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/Zaoqu-Liu/cellGeometry"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>cellGeometry: Quick Start Guide</h1>
                        <h4 data-toc-skip class="author">Zaoqu Liu,
Myles Lewis</h4>
            
            <h4 data-toc-skip class="date">2026-02-03</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/Zaoqu-Liu/cellGeometry/blob/HEAD/vignettes/intro.Rmd" class="external-link"><code>vignettes/intro.Rmd</code></a></small>
      <div class="d-none name"><code>intro.Rmd</code></div>
    </div>

    
    
<div id="geometric-single-cell-deconvolution" class="section level1">
<h1>Geometric single cell deconvolution</h1>
<p><code>cellGeometry</code> has been written for ultrafast
deconvolution of bulk RNA-Seq datasets using a single-cell RNA-Seq
reference dataset in which cell clusters have been defined.</p>
</div>
<div id="installation" class="section level1">
<h1>Installation</h1>
<p>Bioconductor version &gt;=3.20 must be installed first for this
package to install correctly. For full package functionality,
particularly with sparse matrices stored on disc in the h5ad format, we
recommend that the Bioconductor packages zellkonverter, rhdf5 and
HDF5Array must also be installed to be able to read h5ad files. If you
are using Seurat, then it needs to be installed. We also recommend
installing AnnotationHub to enable conversion of ensembl gene ids to
symbols.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="co"># Bioconductor must be installed +/- updated first</span></span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a>BiocManager<span class="sc">::</span><span class="fu">install</span>(<span class="at">version =</span> <span class="st">&quot;3.xx&quot;</span>)  <span class="co"># set to latest version</span></span>
<span id="cb1-3"><a href="#cb1-3" tabindex="-1"></a></span>
<span id="cb1-4"><a href="#cb1-4" tabindex="-1"></a><span class="co"># minimum necessary Bioconductor packages to install cellGeometry package</span></span>
<span id="cb1-5"><a href="#cb1-5" tabindex="-1"></a>BiocManager<span class="sc">::</span><span class="fu">install</span>(<span class="fu">c</span>(<span class="st">&quot;ensembldb&quot;</span>, <span class="st">&quot;DelayedArray&quot;</span>))</span>
<span id="cb1-6"><a href="#cb1-6" tabindex="-1"></a></span>
<span id="cb1-7"><a href="#cb1-7" tabindex="-1"></a><span class="co"># packages needed to read h5ad files</span></span>
<span id="cb1-8"><a href="#cb1-8" tabindex="-1"></a>BiocManager<span class="sc">::</span><span class="fu">install</span>(<span class="fu">c</span>(<span class="st">&quot;zellkonverter&quot;</span>, <span class="st">&quot;rhdf5&quot;</span>, <span class="st">&quot;HDF5Array&quot;</span>))</span>
<span id="cb1-9"><a href="#cb1-9" tabindex="-1"></a></span>
<span id="cb1-10"><a href="#cb1-10" tabindex="-1"></a><span class="co"># optional, if you are using Seurat</span></span>
<span id="cb1-11"><a href="#cb1-11" tabindex="-1"></a><span class="fu">install.packages</span>(<span class="st">&quot;Seurat&quot;</span>)</span>
<span id="cb1-12"><a href="#cb1-12" tabindex="-1"></a></span>
<span id="cb1-13"><a href="#cb1-13" tabindex="-1"></a><span class="co"># package needed to convert ensembl gene ids to symbols</span></span>
<span id="cb1-14"><a href="#cb1-14" tabindex="-1"></a>BiocManager<span class="sc">::</span><span class="fu">install</span>(<span class="st">&quot;AnnotationHub&quot;</span>)</span></code></pre></div>
<p>Install from R-universe (Recommended)</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="fu">install.packages</span>(<span class="st">&quot;cellGeometry&quot;</span>, <span class="at">repos =</span> <span class="st">&quot;https://zaoqu-liu.r-universe.dev&quot;</span>)</span></code></pre></div>
<p>Install from Github</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a>devtools<span class="sc">::</span><span class="fu">install_github</span>(<span class="st">&quot;Zaoqu-Liu/cellGeometry&quot;</span>)</span></code></pre></div>
</div>
<div id="algorithm" class="section level1">
<h1>Algorithm</h1>
<p>The algorithm is performed in two stages:</p>
<ol style="list-style-type: decimal">
<li><p>Optimal gene markers for each cell subclass are identified. In
this part, each gene is considered as a vector in high dimensions with
cell clusters as dimensions.</p></li>
<li><p>The bulk RNA-Seq is deconvoluted by calculating the vector
projection of each bulk RNA-Seq sample against a vector representing
each cell cluster in high dimensional gene marker space using the vector
dot product. In order to adjust for spillover in the vector projection
between cell clusters, a compensation matrix is applied.</p></li>
</ol>
<p>After each of these stages, diagnostics can be performed and either
the gene signature or deconvolution can be updated.</p>
<p><img src="/Users/liuzaoqu/Downloads/cellGeometry/vignettes/workflow.png" class="r-plt" alt="" width="100%" /></p>
</div>
<div id="example-dataset" class="section level1">
<h1>Example dataset</h1>
<div id="h5ad-file" class="section level2">
<h2>h5ad file</h2>
<p>The following example is based on a the Cell Typist dataset (Global)
which contains 329,762 immune cells and is available on the CZ cellxgene
repository, described here: <a
href="https://cellxgene.cziscience.com/collections/62ef75e4-cbea-454e-a0ce-998ec40223d3"
class="uri">https://cellxgene.cziscience.com/collections/62ef75e4-cbea-454e-a0ce-998ec40223d3</a></p>
<p>The h5ad file (2.9 Gb) for the example can be downloaded from CZ
cellxgene repository directly using this link: <a
href="https://datasets.cellxgene.cziscience.com/2ac906a5-9725-4258-8e36-21a9f6c0302a.h5ad"
class="uri">https://datasets.cellxgene.cziscience.com/2ac906a5-9725-4258-8e36-21a9f6c0302a.h5ad</a></p>
<p>First we load the file in HDF5 format so that the full data remains
on disc and only subsets of the data are loaded/processed when necessary
using the HDF5Array and DelayedArray packages.</p>
<p>It is strongly recommended that the file is located on a fast local
drive, preferably an SSD, and not a networked drive. For example, an
analysis of a midsized dataset only took 10 minutes when the h5 file was
stored on a local SSD, compared to 3 hours when the file was stored on a
NAS.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a><span class="fu">library</span>(zellkonverter)</span>
<span id="cb4-2"><a href="#cb4-2" tabindex="-1"></a><span class="fu">library</span>(SingleCellExperiment)</span>
<span id="cb4-3"><a href="#cb4-3" tabindex="-1"></a><span class="fu">library</span>(cellGeometry)</span>
<span id="cb4-4"><a href="#cb4-4" tabindex="-1"></a></span>
<span id="cb4-5"><a href="#cb4-5" tabindex="-1"></a>typist_h5 <span class="ot">&lt;-</span> <span class="fu">readH5AD</span>(<span class="st">&quot;2ac906a5-9725-4258-8e36-21a9f6c0302a.h5ad&quot;</span>,</span>
<span id="cb4-6"><a href="#cb4-6" tabindex="-1"></a>                      <span class="at">use_hdf5 =</span> <span class="cn">TRUE</span>, <span class="at">reader =</span> <span class="st">&quot;R&quot;</span>)</span></code></pre></div>
<p>We extract the main count matrix and cell metadata. cellGeometry
needs rownames on the count matrix.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a>mat <span class="ot">&lt;-</span> typist_h5<span class="sc">@</span>assays<span class="sc">@</span>data<span class="sc">$</span>X</span>
<span id="cb5-2"><a href="#cb5-2" tabindex="-1"></a><span class="fu">rownames</span>(mat) <span class="ot">&lt;-</span> <span class="fu">rownames</span>(typist_h5)</span>
<span id="cb5-3"><a href="#cb5-3" tabindex="-1"></a>meta <span class="ot">&lt;-</span> typist_h5<span class="sc">@</span>colData<span class="sc">@</span>listData</span></code></pre></div>
</div>
<div id="seurat-file" class="section level2">
<h2>Seurat file</h2>
<p>Some users report difficulties with installing zellkonverter which
needs working python libraries. cellGeometry can also be used with
Seurat files although these become progressively slower with larger
datasets as well as needing substantial amounts of RAM, so for datasets
&gt;1M cells we recommend persevering with zellkonverter and the h5ad
format since it is much faster. We include example code for loading a
Seurat file below as an alternative to h5ad.</p>
<p>At time of writing the rds file (2.9 Gb) in Seurat format can be
downloaded from CZ cellxgene repository directly using this link: <a
href="https://datasets.cellxgene.cziscience.com/2ac906a5-9725-4258-8e36-21a9f6c0302a.rds"
class="uri">https://datasets.cellxgene.cziscience.com/2ac906a5-9725-4258-8e36-21a9f6c0302a.rds</a></p>
<p>CZ cellxgene state that Seurat support will end after Dec 2024.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a><span class="fu">library</span>(Seurat)</span>
<span id="cb6-2"><a href="#cb6-2" tabindex="-1"></a>typist <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">&quot;2ac906a5-9725-4258-8e36-21a9f6c0302a.rds&quot;</span>)  <span class="co"># 15.5 GB in memory</span></span>
<span id="cb6-3"><a href="#cb6-3" tabindex="-1"></a></span>
<span id="cb6-4"><a href="#cb6-4" tabindex="-1"></a>mat <span class="ot">&lt;-</span> typist<span class="sc">@</span>assays<span class="sc">$</span>RNA<span class="sc">$</span>counts</span>
<span id="cb6-5"><a href="#cb6-5" tabindex="-1"></a>meta <span class="ot">&lt;-</span> typist<span class="sc">@</span>meta.data</span></code></pre></div>
</div>
</div>
<div id="obtain-gene-signatures" class="section level1">
<h1>Obtain gene signatures</h1>
<div id="extract-cell-clusters" class="section level2">
<h2>Extract cell clusters</h2>
<p>We first check cell cluster subclasses. Then we extract a vector
which contains the subclass cluster for each cell and a 2nd vector for
broader cell groups. We restrict the dataset to blood so that we can
deconvolute blood bulk RNA-Seq data later (since tissue-based cells
should not be found in peripheral blood samples). This reduces the
analysis to 27,602 cells out of the total 329,762 cells and reduces the
number of subclasses from 43 to 27. This is entirely optional and the
algorithm can easily run the full analysis.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a><span class="fu">table</span>(meta<span class="sc">$</span>Majority_voting_CellTypist)</span>
<span id="cb7-2"><a href="#cb7-2" tabindex="-1"></a></span>
<span id="cb7-3"><a href="#cb7-3" tabindex="-1"></a>subcl <span class="ot">&lt;-</span> meta<span class="sc">$</span>Majority_voting_CellTypist</span>
<span id="cb7-4"><a href="#cb7-4" tabindex="-1"></a>cellgrp <span class="ot">&lt;-</span> meta<span class="sc">$</span>Majority_voting_CellTypist_high</span>
<span id="cb7-5"><a href="#cb7-5" tabindex="-1"></a></span>
<span id="cb7-6"><a href="#cb7-6" tabindex="-1"></a><span class="co"># reduce dataset to only blood (optional)</span></span>
<span id="cb7-7"><a href="#cb7-7" tabindex="-1"></a>subcl[meta<span class="sc">$</span>tissue <span class="sc">!=</span> <span class="st">&quot;blood&quot;</span>] <span class="ot">&lt;-</span> <span class="cn">NA</span></span>
<span id="cb7-8"><a href="#cb7-8" tabindex="-1"></a>cellgrp[meta<span class="sc">$</span>tissue <span class="sc">!=</span> <span class="st">&quot;blood&quot;</span>] <span class="ot">&lt;-</span> <span class="cn">NA</span></span></code></pre></div>
</div>
<div id="create-gene-signature" class="section level2">
<h2>Create gene signature</h2>
<p>We then run the 1st stage of cellGeometry which generates mean gene
expression for each cell cluster (this is the slowest part). Then the
best cell cluster and cell group gene markers are identified.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" tabindex="-1"></a>mk <span class="ot">&lt;-</span> <span class="fu">cellMarkers</span>(mat, <span class="at">subclass =</span> subcl, <span class="at">cellgroup =</span> cellgrp,</span>
<span id="cb8-2"><a href="#cb8-2" tabindex="-1"></a>                  <span class="at">dual_mean =</span> <span class="cn">TRUE</span>, <span class="at">cores =</span> <span class="dv">2</span>)</span></code></pre></div>
<p>The <code>dual_mean</code> argument only needs to be set for the
purpose of the simulation later. Most users do not need to set this. It
calculates both the standard mean gene expression, which is
mean(log<sub>2</sub>(counts +1)), as well as the arithmetic mean of the
(unlogged) counts.</p>
<p>The derivation of mean gene expression for each cluster and cell
group is the slowest part. If you are on linux or mac, this can be sped
up using parallelisation by setting <code>cores = 2</code> or more. Note
that this can increase memory requirements dramatically unless HFD5 is
used. For this particular dataset which is moderate in size, we find
significant speed up with 4-8 cores (64 Gb machine). For very large
datasets (&gt;1M cells) if the sc data is kept on disc via HFD5 then
many cores can be used. But if the data or subsets of it have to be
loaded into memory then we typically apportion around 16 Gb per core
(e.g. 3 cores on a 64 Gb machine). So the limit on cores depends on the
size of the single-cell data, available RAM and whether HFD5 is
used.</p>
<p>Windows users can invoke parallelisation using the future backend and
setting up a multisession plan.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" tabindex="-1"></a><span class="co"># example code using future for parallelisation on windows</span></span>
<span id="cb9-2"><a href="#cb9-2" tabindex="-1"></a><span class="fu">library</span>(future)</span>
<span id="cb9-3"><a href="#cb9-3" tabindex="-1"></a><span class="fu">plan</span>(multisession, <span class="at">workers =</span> <span class="dv">4</span>)</span>
<span id="cb9-4"><a href="#cb9-4" tabindex="-1"></a></span>
<span id="cb9-5"><a href="#cb9-5" tabindex="-1"></a>mk <span class="ot">&lt;-</span> <span class="fu">cellMarkers</span>(mat, <span class="at">subclass =</span> subcl, <span class="at">cellgroup =</span> cellgrp,</span>
<span id="cb9-6"><a href="#cb9-6" tabindex="-1"></a>                  <span class="at">use_future =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<p>We have not specified a bulk RNA-Seq dataset at this stage as this
example is based on simulation alone. However, if you have a bulk
RNA-Seq dataset it is helpful to specify it during the first call to
<code>cellMarkers()</code>. It is only used for its rownames to identify
genes that overlap between the 2 datasets. The marker signature can be
updated later for different bulk datasets using
<code>updateMarkers()</code> (see below).</p>
<p>We convert the ensembl ids in the cellMarkers object using the
built-in function <code>gene2symbol()</code>. This needs an ensembl
database to be loaded.</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" tabindex="-1"></a><span class="fu">library</span>(AnnotationHub)</span>
<span id="cb10-2"><a href="#cb10-2" tabindex="-1"></a>ah <span class="ot">&lt;-</span> <span class="fu">AnnotationHub</span>()</span>
<span id="cb10-3"><a href="#cb10-3" tabindex="-1"></a>ensDb_v110 <span class="ot">&lt;-</span> ah[[<span class="st">&quot;AH113665&quot;</span>]]</span>
<span id="cb10-4"><a href="#cb10-4" tabindex="-1"></a>mk <span class="ot">&lt;-</span> <span class="fu">gene2symbol</span>(mk, ensDb_v110)</span></code></pre></div>
</div>
<div id="visualise-gene-signature" class="section level2">
<h2>Visualise gene signature</h2>
<p>The signature gene matrix can be displayed as follows.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" tabindex="-1"></a><span class="fu">signature_heatmap</span>(mk)  <span class="co"># visualise whole signature</span></span>
<span id="cb11-2"><a href="#cb11-2" tabindex="-1"></a><span class="fu">signature_heatmap</span>(mk, <span class="at">top =</span> <span class="dv">5</span>)  <span class="co"># show top 5 genes for each subclass</span></span></code></pre></div>
<p><img src="/Users/liuzaoqu/Downloads/cellGeometry/vignettes/typist_sig.png" class="r-plt" alt="" width="90%" /></p>
<p>The default signature heatmap shows the gene signature after the
noise filter has been applied. To see the raw gene expression heatmap
call <code>signature_heatmap()</code> with
<code>use_filter = FALSE</code>.</p>
<p>The spillover heatmap between cell clusters can also be
visualised.</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" tabindex="-1"></a><span class="fu">spillover_heatmap</span>(mk)</span></code></pre></div>
<p><img src="/Users/liuzaoqu/Downloads/cellGeometry/vignettes/typist_spillover.png" class="r-plt" alt="" width="80%" /></p>
<p>This heatmap as well as the signature heatmap reveals that some cell
subclasses ‘spillover’ too strongly into other cell subclasses. In other
words some cell types are too similar - perhaps one is really a closely
related subset of the other. Here we see that Helper T cells are the
most affected and their signature is similar to Tcm/Naive helper T
cells.</p>
</div>
<div id="signature-diagnostics" class="section level2">
<h2>Signature diagnostics</h2>
<p>When cell clusters exhibit spillover, it is generally because of cell
types being too similar. Cosine similarity is an ideal method for
detecting this as it explicitly measures the angle between the vectors
for each cell subclass as defined by in gene space.
<code>cos_similarity()</code> generates a cosine similarity matrix for
the cell subclasses, and this can be visualised using a heatmap.</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" tabindex="-1"></a>cs <span class="ot">&lt;-</span> <span class="fu">cos_similarity</span>(mk)</span>
<span id="cb13-2"><a href="#cb13-2" tabindex="-1"></a>ComplexHeatmap<span class="sc">::</span><span class="fu">Heatmap</span>(cs)</span></code></pre></div>
<p><img src="/Users/liuzaoqu/Downloads/cellGeometry/vignettes/typist_cos_sim.png" class="r-plt" alt="" width="80%" /></p>
<p>The cosine similarity matrix is easily be converted to the angle
between the vectors for each cell subclass using <code>acos(cs)</code>.
The function <code>rank_angle()</code> can be used to list those pairs
of subclasses whose vectors are close together in terms of the angle
between them.</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" tabindex="-1"></a><span class="fu">rank_angle</span>(cs)</span></code></pre></div>
<p>This plus the cosine similarity heatmap confirms that the vector for
Helper T cells is rather similar to the vector for Tcm/Naive helper T
cells, which is not surprising as these cell types overlap.</p>
<p>The <code>diagnose()</code> function provides more thorough
diagnostics on the gene signature. It can be used to identify which cell
clusters are problematic. Helper T cells in particular do not have any
top ranked markers of their own since they are largely identical to
Tcm/Naive helper T cells in terms of their gene expression.</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" tabindex="-1"></a><span class="fu">diagnose</span>(mk)</span></code></pre></div>
<p>Below we update the cellMarkers object to remove 2 cell clusters
which overlap with other cell clusters and are therefore likely to be
difficult to deconvolute well if applied to real world bulk RNA-Seq.
(For the simulation it actually does not matter whether these are
removed or not since the algorithm is able to differentiate them
sufficiently.)</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" tabindex="-1"></a>mk <span class="ot">&lt;-</span> <span class="fu">updateMarkers</span>(mk,</span>
<span id="cb16-2"><a href="#cb16-2" tabindex="-1"></a>                    <span class="at">remove_subclass =</span> <span class="fu">c</span>(<span class="st">&quot;Helper T cells&quot;</span>, <span class="st">&quot;Cytotoxic T cells&quot;</span>))</span></code></pre></div>
</div>
<div id="refine-gene-signature" class="section level2">
<h2>Refine gene signature</h2>
<p>The following arguments are available in both
<code>cellMarkers()</code> or <code>updateMarkers()</code>. They can
change and improve which genes are selected for the gene signature as
well as changing the amount of noise reduction in the gene signature if
the filter is applied during deconvolution by
<code>deconvolute()</code>.</p>
<table>
<colgroup>
<col width="17%" />
<col width="82%" />
</colgroup>
<tbody>
<tr class="odd">
<td align="left"><code>nsubclass</code></td>
<td align="left">Number of genes to select for each single cell
subclass. Either a single number or a vector with the number of genes
for each subclass.</td>
</tr>
<tr class="even">
<td align="left"><code>ngroup</code></td>
<td align="left">Number of genes to select for each cell group. Either a
single number or a vector with the number of genes for each group.</td>
</tr>
<tr class="odd">
<td align="left"><code>expfilter</code></td>
<td align="left">Genes whose maximum mean expression on log2 scale per
cell type are below this value are removed and not considered for the
signature.</td>
</tr>
<tr class="even">
<td align="left"><code>noisefilter</code></td>
<td align="left">Sets an upper bound for <code>noisefraction</code>
cut-off below which gene expression is set to 0. Essentially gene
expression above this level must be retained in the signature. Setting
this higher can allow more suppression via <code>noisefraction</code>
and can favour more highly expressed genes.</td>
</tr>
<tr class="odd">
<td align="left"><code>noisefraction</code></td>
<td align="left">Numeric value from 0-1 (default 0.25). Higher values
mean more noise suppression. Maximum mean log2 gene expression across
cell types is calculated and values in celltypes below this fraction are
set to 0. Set in conjunction with <code>noisefilter.</code> Note: if
this is set too high (too close to 1), it can have a deleterious effect
on deconvolution.</td>
</tr>
</tbody>
</table>
<p><code>nsubclass</code> and <code>ngroup</code> simply define how many
genes are picked for each subclass or group. If some subclasses do not
have enough specific genes, lowering <code>expfilter</code> is probably
the first parameter to try. However as <code>expfilter</code> approaches
zero, lots of genes with very low expression but which appear to be
highly specific to a subclass (i.e. only expressed in that subclass with
zero expression in all other cell clusters) will be selected. Although
these might perform well in simulations, our experience is that they are
poor markers in real bulk RNA-Seq.</p>
<p>In addition users have full control over editing the gene list for
the subclass signature using the <code>add_gene</code> and
<code>remove_gene</code> arguments in <code>updateMarkers()</code>.
Similarly <code>add_groupgene</code> and <code>remove_groupgene</code>
can be used to edit the gene list for the broader cell group
signatures.</p>
</div>
</div>
<div id="deconvolution" class="section level1">
<h1>Deconvolution</h1>
<div id="simulated-pseudo-bulk" class="section level2">
<h2>Simulated pseudo-bulk</h2>
<p>We can generate pseudo-bulk to test the deconvolution using the
following commands. Here <code>generate_samples()</code> makes 25
samples with random cell counts, <code>sim_counts</code>. The
simulate_bulk() function operates in 2 modes. In the first mode, the
average gene expression for each cell cluster is extracted from the
cellMarkers object and used to generate the pseudo-bulk totals. In the
2nd mode (see below) the original single-cell count data is sampled.</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" tabindex="-1"></a><span class="co"># simulated bulk</span></span>
<span id="cb17-2"><a href="#cb17-2" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">3</span>)</span>
<span id="cb17-3"><a href="#cb17-3" tabindex="-1"></a>sim_counts <span class="ot">&lt;-</span> <span class="fu">generate_samples</span>(mk, <span class="dv">25</span>)</span>
<span id="cb17-4"><a href="#cb17-4" tabindex="-1"></a>sim_percent <span class="ot">&lt;-</span> sim_counts <span class="sc">/</span> <span class="fu">rowSums</span>(sim_counts) <span class="sc">*</span> <span class="dv">100</span></span>
<span id="cb17-5"><a href="#cb17-5" tabindex="-1"></a>sim_pseudo <span class="ot">&lt;-</span> <span class="fu">simulate_bulk</span>(mk, sim_counts)</span></code></pre></div>
<p>Deconvolution itself is performed as a 2nd function
<code>deconvolute()</code>. The <code>plot_set()</code> function can be
used to plot the results. The <code>metric_set()</code> function
generates a table of results.</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" tabindex="-1"></a><span class="co"># mode 1: (perfect deconvolution)</span></span>
<span id="cb18-2"><a href="#cb18-2" tabindex="-1"></a>fit <span class="ot">&lt;-</span> <span class="fu">deconvolute</span>(mk, sim_pseudo,</span>
<span id="cb18-3"><a href="#cb18-3" tabindex="-1"></a>                   <span class="at">use_filter =</span> <span class="cn">FALSE</span>)</span>
<span id="cb18-4"><a href="#cb18-4" tabindex="-1"></a><span class="fu">plot_set</span>(sim_counts, fit<span class="sc">$</span>subclass<span class="sc">$</span>output)</span>
<span id="cb18-5"><a href="#cb18-5" tabindex="-1"></a><span class="fu">plot_set</span>(sim_percent, fit<span class="sc">$</span>subclass<span class="sc">$</span>percent)</span>
<span id="cb18-6"><a href="#cb18-6" tabindex="-1"></a></span>
<span id="cb18-7"><a href="#cb18-7" tabindex="-1"></a><span class="fu">metric_set</span>(sim_percent, fit<span class="sc">$</span>subclass<span class="sc">$</span>percent)  <span class="co"># table of results</span></span></code></pre></div>
</div>
<div id="sampled-pseudo-bulk" class="section level2">
<h2>Sampled pseudo-bulk</h2>
<p>In the 2nd mode, the original scRNA-Seq count dataset is sampled. The
sampling rate of the actual cell counts in <code>sim_counts</code> can
be increased by setting <code>times</code>. Cells are sampled with
replacement. The desired cell counts are simply multiplied by
<code>times</code> prior to sampling. By default, sampling is performed
using the Dirichlet distribution as this gives true random sampling in
comparison to uniform sampling. When cells are oversampled uniformly, in
the limit the summed gene expression tends to the arithmetic mean of the
subclass x sample frequency (which is a much easier deconvolution
problem compared to Dirichlet sampling). This can be demonstrated by
using uniform sampling and steadily increasing <code>times</code> from 1
to 30-100 or more. This improves the deconvolution as the sum of the
gene counts per sampled cell approaches the arithmetic mean of gene
counts for each cell cluster.</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" tabindex="-1"></a><span class="co"># mode 2: sample from original sc count matrix</span></span>
<span id="cb19-2"><a href="#cb19-2" tabindex="-1"></a><span class="co"># 1.43 mins (Intel); 45 secs (ARM)</span></span>
<span id="cb19-3"><a href="#cb19-3" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">99</span>)</span>
<span id="cb19-4"><a href="#cb19-4" tabindex="-1"></a>times <span class="ot">&lt;-</span> <span class="dv">1</span>  <span class="co"># can be increased</span></span>
<span id="cb19-5"><a href="#cb19-5" tabindex="-1"></a>sim_sampled <span class="ot">&lt;-</span> <span class="fu">simulate_bulk</span>(mat, sim_counts, subcl, <span class="at">times =</span> times)</span>
<span id="cb19-6"><a href="#cb19-6" tabindex="-1"></a></span>
<span id="cb19-7"><a href="#cb19-7" tabindex="-1"></a><span class="co"># fix rownames</span></span>
<span id="cb19-8"><a href="#cb19-8" tabindex="-1"></a><span class="fu">rownames</span>(sim_sampled) <span class="ot">&lt;-</span> <span class="fu">gene2symbol</span>(<span class="fu">rownames</span>(sim_sampled), ensDb_v110)</span>
<span id="cb19-9"><a href="#cb19-9" tabindex="-1"></a></span>
<span id="cb19-10"><a href="#cb19-10" tabindex="-1"></a><span class="co"># near optimal deconvolution of counts sampled from the original scRNA-Seq</span></span>
<span id="cb19-11"><a href="#cb19-11" tabindex="-1"></a>fit2 <span class="ot">&lt;-</span> <span class="fu">deconvolute</span>(mk, sim_sampled,</span>
<span id="cb19-12"><a href="#cb19-12" tabindex="-1"></a>                    <span class="at">use_filter =</span> <span class="cn">FALSE</span>, <span class="at">arith_mean =</span> <span class="cn">TRUE</span>)</span>
<span id="cb19-13"><a href="#cb19-13" tabindex="-1"></a></span>
<span id="cb19-14"><a href="#cb19-14" tabindex="-1"></a><span class="co"># plot results</span></span>
<span id="cb19-15"><a href="#cb19-15" tabindex="-1"></a><span class="fu">plot_set</span>(sim_counts, fit2<span class="sc">$</span>subclass<span class="sc">$</span>output <span class="sc">/</span> times)  <span class="co"># adjust for oversampling using `times`</span></span>
<span id="cb19-16"><a href="#cb19-16" tabindex="-1"></a><span class="fu">plot_set</span>(sim_percent, fit2<span class="sc">$</span>subclass<span class="sc">$</span>percent)</span>
<span id="cb19-17"><a href="#cb19-17" tabindex="-1"></a></span>
<span id="cb19-18"><a href="#cb19-18" tabindex="-1"></a><span class="fu">metric_set</span>(sim_percent, fit2<span class="sc">$</span>subclass<span class="sc">$</span>percent)</span></code></pre></div>
<p><img src="/Users/liuzaoqu/Downloads/cellGeometry/vignettes/typist_sim.png" class="r-plt" alt="" width="90%" /></p>
<p>Results are returned in <code>fit$subclass$output</code> for standard
output (theoretically in numbers of cells),
<code>fit$subclass$percent</code> for output converted to percentage of
total cells for each sample. The group analysis is returned in a similar
vein in <code>fit$group$output</code> and
<code>fit$group$percent</code>.</p>
</div>
<div id="real-bulk-rna-seq-settings" class="section level2">
<h2>Real bulk RNA-Seq settings</h2>
<p>We recommend that bulk RNA-Seq sample data is provided as raw counts.
Log<sub>2</sub> transformed data can be inputted, but it is important
that the transformed data is <em>not</em> offset, i.e. the algorithm
expects that zero in the bulk really means zero expression. Some
transformation methods, e.g. variance stabilising transform (VST), mean
that there is an offset so that zero counts equates to a positive
number. This typed of transformed data cannot be used.</p>
<p>With a real bulk RNA-Seq dataset the first step is to ensure that
only genes which are present in the bulk data are used for the gene
signatures. This is achieved using the function
<code>updateMarkers()</code>.</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" tabindex="-1"></a>mk <span class="ot">&lt;-</span> <span class="fu">updateMarkers</span>(mk, <span class="at">bulkdata =</span> my_bulk_matrix)</span></code></pre></div>
<p>Alternatively bulk data can be supplied with the first call to
<code>cellMarkers()</code>. <code>updateMarkers()</code> can also be
used to rapidly update the marker object <code>mk</code> with new
settings, e.g. to alter the number of genes used per subclass or
manually edit (add/remove) the marker genes. Note that if some signature
genes are missing from the bulk data, <code>deconvolute()</code> will
stop with an error message that some signature genes are missing.</p>
<p>Note that the settings shown above are mathematically ideal for
simulated bulk data. In reality, we expect the scRNA-Seq signature to
differ from real-world bulk RNA-Seq due to differences in chemistry and
the amplification step required by single-cell sequencing. So we
recommend the default settings for real-world bulk data when calling
<code>deconvolute()</code>. The main settings to choose here are:</p>
<table>
<colgroup>
<col width="18%" />
<col width="13%" />
<col width="14%" />
<col width="52%" />
</colgroup>
<thead>
<tr class="header">
<th align="left">Parameter</th>
<th align="left">Optimal simulation</th>
<th align="left">Real-world recommended</th>
<th align="left">Purpose</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left"><code>count_space</code></td>
<td align="left">TRUE</td>
<td align="left">TRUE</td>
<td align="left">Toggles whether the deconvolution is performed in count
space (exponential space) or in log<sub>2</sub> space.</td>
</tr>
<tr class="even">
<td align="left"><code>weight_method</code></td>
<td align="left">“equal”</td>
<td align="left">“equal”</td>
<td align="left">Scaling is applied so that each gene in the gene
signature has equal weight.</td>
</tr>
<tr class="odd">
<td align="left"><code>convert_bulk</code></td>
<td align="left">FALSE</td>
<td align="left">FALSE</td>
<td align="left">Converts the bulk RNA-Seq onto scRNA-Seq data scale
based on a reference dataset in which both bulk and scRNA-Seq were
performed simultaneously.</td>
</tr>
<tr class="even">
<td align="left"><code>use_filter</code></td>
<td align="left"><strong>FALSE</strong></td>
<td align="left">TRUE</td>
<td align="left">Reduces the overall amount of spillover &amp;
compensation by denoising the gene signatures.</td>
</tr>
<tr class="odd">
<td align="left"><code>arith_mean</code></td>
<td align="left"><strong>TRUE</strong></td>
<td align="left">FALSE</td>
<td align="left">Whether to use arithmetic mean of counts or the
standard log<sub>2</sub> mean.</td>
</tr>
</tbody>
</table>
<p>We recommend <code>use_filter</code> is always on for real-world bulk
analysis as it reduces the compensation burden particularly for genes
where non-specific low level expression is present in some cell clusters
which can occur even for good markers. We suggest
<code>deconvolute()</code> is used in one of 2 modes: either
<code>convert_bulk = TRUE</code> or <code>count_space = TRUE</code>, but
<strong>not both</strong>.</p>
<p>There is also a powerful function <code>tune_deconv()</code> which
allows users to tune any of the parameters available in
<code>updateMarkers()</code> based on a bulk reference dataset. The
simulated pseudo-bulk data can be used for this purpose, but real bulk
RNA-Seq would be better (more realistic and better for tuning).</p>
<p>Also, 2 scRNA-Seq datasets can be merged using the function
<code>mergeMarkers()</code>. This merges the <code>cellMarkers</code>
objects derived from each single cell dataset. One dataset is defined as
reference, and the 2nd dataset is merged into it after adjustment for
its overall distribution based on quantile mapping.</p>
</div>
</div>
<div id="tuning-deconvolution" class="section level1">
<h1>Tuning deconvolution</h1>
<p>The function <code>tune_deconv()</code> can be used with the sampled
pseudo-bulk matrix <code>sim_sampled</code> which we generated earlier
to see the effect of tuning parameters for deconvolution. Note there are
important mathematical caveats to understand when using tuning. Below we
tune over a range of <code>nsubclass</code> values from 5 up to 1000,
vary the low expression filter <code>expfilter</code> from 0.05 up to
1.5, and also tune <code>weight_method</code> to either “none” or
“equal”. But any argument which can be passed to
<code>updateMarkers()</code> can be tuned. Arguments which are not in
the tuning list specified through the argument <code>grid</code> such as
<code>arith_mean</code> and <code>use_filter</code> in the example below
are applied constantly. Parallelisation is available on unix systems but
not on windows.</p>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" tabindex="-1"></a>res <span class="ot">&lt;-</span> <span class="fu">tune_deconv</span>(mk, sim_sampled, sim_counts <span class="sc">*</span> times,</span>
<span id="cb21-2"><a href="#cb21-2" tabindex="-1"></a>                   <span class="at">grid =</span> <span class="fu">list</span>(<span class="at">nsubclass =</span> <span class="fu">c</span>(<span class="dv">5</span>, <span class="dv">10</span>, <span class="dv">15</span>, <span class="dv">25</span>, <span class="dv">50</span>, <span class="dv">100</span>, <span class="dv">200</span>, <span class="dv">500</span>, <span class="dv">1000</span>),</span>
<span id="cb21-3"><a href="#cb21-3" tabindex="-1"></a>                               <span class="at">expfilter =</span> <span class="fu">c</span>(<span class="fl">0.05</span>, <span class="fl">0.1</span>, <span class="fl">0.25</span>, <span class="fl">0.5</span>, <span class="fl">0.75</span>, <span class="dv">1</span>, <span class="fl">1.5</span>),</span>
<span id="cb21-4"><a href="#cb21-4" tabindex="-1"></a>                               <span class="at">weight_method =</span> <span class="fu">c</span>(<span class="st">&quot;none&quot;</span>, <span class="st">&quot;equal&quot;</span>)),</span>
<span id="cb21-5"><a href="#cb21-5" tabindex="-1"></a>                   <span class="at">arith_mean =</span> <span class="cn">TRUE</span>,</span>
<span id="cb21-6"><a href="#cb21-6" tabindex="-1"></a>                   <span class="at">use_filter =</span> <span class="cn">FALSE</span>,</span>
<span id="cb21-7"><a href="#cb21-7" tabindex="-1"></a>                   <span class="at">cores =</span> <span class="dv">8</span>)</span>
<span id="cb21-8"><a href="#cb21-8" tabindex="-1"></a></span>
<span id="cb21-9"><a href="#cb21-9" tabindex="-1"></a><span class="do">## Tuning parameters: nsubclass, expfilter, weight_method</span></span>
<span id="cb21-10"><a href="#cb21-10" tabindex="-1"></a><span class="do">##   |==========================================================| 100%, Elapsed 00:07</span></span>
<span id="cb21-11"><a href="#cb21-11" tabindex="-1"></a><span class="do">## Best tune:</span></span>
<span id="cb21-12"><a href="#cb21-12" tabindex="-1"></a><span class="do">##   nsubclass  expfilter  weight_method  mean.RMSE</span></span>
<span id="cb21-13"><a href="#cb21-13" tabindex="-1"></a><span class="do">##        1000       0.25          equal      35.21</span></span></code></pre></div>
<p>When tuning it is important not to forget the effect of
<code>times</code> in <code>simulate_bulk()</code>. That is why the
supplied true cell counts matrix <code>sim_counts</code> is multiplied
by the oversampling factor <code>times</code> in the above example.</p>
<p>The results of tuning can be plotted using the function
<code>plot_tune()</code>.</p>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" tabindex="-1"></a><span class="fu">plot_tune</span>(res, <span class="at">xvar =</span> <span class="st">&quot;nsubclass&quot;</span>, <span class="at">group =</span> <span class="st">&quot;weight_method&quot;</span>)</span>
<span id="cb22-2"><a href="#cb22-2" tabindex="-1"></a><span class="fu">plot_tune</span>(res, <span class="at">xvar =</span> <span class="st">&quot;nsubclass&quot;</span>, <span class="at">group =</span> <span class="st">&quot;expfilter&quot;</span>)</span>
<span id="cb22-3"><a href="#cb22-3" tabindex="-1"></a><span class="fu">plot_tune</span>(res, <span class="at">xvar =</span> <span class="st">&quot;expfilter&quot;</span>, <span class="at">group =</span> <span class="st">&quot;weight_method&quot;</span>)</span>
<span id="cb22-4"><a href="#cb22-4" tabindex="-1"></a><span class="fu">plot_tune</span>(res, <span class="at">xvar =</span> <span class="st">&quot;expfilter&quot;</span>, <span class="at">group =</span> <span class="st">&quot;nsubclass&quot;</span>)</span></code></pre></div>
<p><img src="/Users/liuzaoqu/Downloads/cellGeometry/vignettes/typist_tune.png" class="r-plt" alt="" width="90%" /></p>
<p>These plots show that <code>weight_method = "equal"</code> is better
than <code>"none"</code> and that <code>expfilter</code> for this single
cell dataset seems to be optimal at around 0.25. In this simulation the
higher <code>nsubclass</code> is the better, with deconvolution best
with 500-1000 genes per subclass. In reality we find that this is
unlikely to be true for real-world bulk RNA-Seq as noise caused by
differences in sequencing chemistry mean that larger numbers of genes in
the signature add increasing noise, so there is likely to be a
sweetspot. The default <code>nsubclass</code> is 25 which has been
chosen as a plausible compromise which should work in most
situations.</p>
<div id="simulating-noise" class="section level2">
<h2>Simulating noise</h2>
<p>The package provides several noise adding functions which can be
applied to the sampled simulation dataset.</p>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" tabindex="-1"></a><span class="co"># simple gaussian noise applied to counts</span></span>
<span id="cb23-2"><a href="#cb23-2" tabindex="-1"></a>sim_noise <span class="ot">&lt;-</span> <span class="fu">add_noise</span>(sim_sampled)</span>
<span id="cb23-3"><a href="#cb23-3" tabindex="-1"></a></span>
<span id="cb23-4"><a href="#cb23-4" tabindex="-1"></a><span class="co"># noise applied to log2 counts</span></span>
<span id="cb23-5"><a href="#cb23-5" tabindex="-1"></a>sim_noise <span class="ot">&lt;-</span> <span class="fu">log_noise</span>(sim_sampled)</span>
<span id="cb23-6"><a href="#cb23-6" tabindex="-1"></a></span>
<span id="cb23-7"><a href="#cb23-7" tabindex="-1"></a><span class="co"># noise applied to sqrt transformed counts</span></span>
<span id="cb23-8"><a href="#cb23-8" tabindex="-1"></a>sim_noise <span class="ot">&lt;-</span> <span class="fu">sqrt_noise</span>(sim_sampled)</span>
<span id="cb23-9"><a href="#cb23-9" tabindex="-1"></a></span>
<span id="cb23-10"><a href="#cb23-10" tabindex="-1"></a><span class="co"># whole genes are scaled up/down by a random amount</span></span>
<span id="cb23-11"><a href="#cb23-11" tabindex="-1"></a><span class="co"># this simulates differences in chemistry</span></span>
<span id="cb23-12"><a href="#cb23-12" tabindex="-1"></a>sim_noise <span class="ot">&lt;-</span> <span class="fu">shift_noise</span>(sim_sampled)</span></code></pre></div>
<p>Tuning can then be tested on the simulation data with added noise to
see how tolerant the deconvolution parameters are to different types of
noise.</p>
<div class="sourceCode" id="cb24"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" tabindex="-1"></a>res2 <span class="ot">&lt;-</span> <span class="fu">tune_deconv</span>(mk, sim_noise, sim_counts,</span>
<span id="cb24-2"><a href="#cb24-2" tabindex="-1"></a>                   <span class="at">grid =</span> <span class="fu">list</span>(<span class="at">nsubclass =</span> <span class="fu">c</span>(<span class="dv">5</span>, <span class="dv">10</span>, <span class="dv">15</span>, <span class="dv">25</span>, <span class="dv">50</span>, <span class="dv">100</span>, <span class="dv">200</span>, <span class="dv">500</span>, <span class="dv">1000</span>),</span>
<span id="cb24-3"><a href="#cb24-3" tabindex="-1"></a>                               <span class="at">expfilter =</span> <span class="fu">c</span>(<span class="fl">0.05</span>, <span class="fl">0.1</span>, <span class="fl">0.25</span>, <span class="fl">0.5</span>, <span class="fl">0.75</span>, <span class="dv">1</span>, <span class="fl">1.5</span>),</span>
<span id="cb24-4"><a href="#cb24-4" tabindex="-1"></a>                               <span class="at">weight_method =</span> <span class="fu">c</span>(<span class="st">&quot;none&quot;</span>, <span class="st">&quot;equal&quot;</span>)),</span>
<span id="cb24-5"><a href="#cb24-5" tabindex="-1"></a>                   <span class="at">arith_mean =</span> <span class="cn">TRUE</span>,</span>
<span id="cb24-6"><a href="#cb24-6" tabindex="-1"></a>                   <span class="at">use_filter =</span> <span class="cn">FALSE</span>,</span>
<span id="cb24-7"><a href="#cb24-7" tabindex="-1"></a>                   <span class="at">cores =</span> <span class="dv">8</span>)</span></code></pre></div>
<p>We find that the addition of simple noise to the simulation dataset
tends to prefer higher <code>nsubclass</code> as this averages out the
noise across more markers. <code>shift_noise</code> which simulates
differences in chemistry tends to show that the low expression filter
<code>expfilter</code> becomes increasingly important and that there is
a maximum limit to <code>nsubclass</code>.</p>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by <a href="https://github.com/Zaoqu-Liu" class="external-link">Zaoqu Liu</a> and <a href="https://github.com/myles-lewis" class="external-link">Myles Lewis</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

    </footer></div>





  </body></html>
