<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Quantile mapping function between two scRNA-Seq datasets — quantile_map • cellGeometry</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Quantile mapping function between two scRNA-Seq datasets — quantile_map"><meta name="description" content="Quantile mapping to combine two scRNA-Seq datasets based on mapping either
the distribution of mean log2+1 gene expression in cell clusters to the
distribution of the 2nd dataset, or mapping the quantiles of one matrix of
gene expression (with genes in rows) to another."><meta property="og:description" content="Quantile mapping to combine two scRNA-Seq datasets based on mapping either
the distribution of mean log2+1 gene expression in cell clusters to the
distribution of the 2nd dataset, or mapping the quantiles of one matrix of
gene expression (with genes in rows) to another."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">cellGeometry</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.5.8</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/intro.html">Get Started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/algorithm.html">Mathematical Foundations</a></li>
    <li><a class="dropdown-item" href="../articles/visualization.html">Visualization Guide</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><a class="dropdown-item" href="../articles/index.html">All articles</a></li>
  </ul></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-tutorials" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Tutorials</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-tutorials"><li><a class="dropdown-item" href="../articles/intro.html">Quick Start Guide</a></li>
    <li><a class="dropdown-item" href="../articles/real_world.html">Real-World Application</a></li>
    <li><a class="dropdown-item" href="../articles/brain_atlas.html">Large-Scale: Brain Atlas</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/Zaoqu-Liu/cellGeometry"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Quantile mapping function between two scRNA-Seq datasets</h1>
      <small class="dont-index">Source: <a href="https://github.com/Zaoqu-Liu/cellGeometry/blob/HEAD/R/quantile_map.R" class="external-link"><code>R/quantile_map.R</code></a></small>
      <div class="d-none name"><code>quantile_map.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Quantile mapping to combine two scRNA-Seq datasets based on mapping either
the distribution of mean log2+1 gene expression in cell clusters to the
distribution of the 2nd dataset, or mapping the quantiles of one matrix of
gene expression (with genes in rows) to another.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">quantile_map</span><span class="op">(</span></span>
<span>  <span class="va">x</span>,</span>
<span>  <span class="va">y</span>,</span>
<span>  n <span class="op">=</span> <span class="fl">10000</span>,</span>
<span>  remove_noncoding <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  remove_zeros <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  smooth <span class="op">=</span> <span class="st">"loess"</span>,</span>
<span>  span <span class="op">=</span> <span class="fl">0.15</span>,</span>
<span>  knots <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.25</span>, <span class="fl">0.75</span>, <span class="fl">0.85</span>, <span class="fl">0.95</span>, <span class="fl">0.97</span>, <span class="fl">0.99</span>, <span class="fl">0.999</span><span class="op">)</span>,</span>
<span>  respace <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  silent <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-x">x<a class="anchor" aria-label="anchor" href="#arg-x"></a></dt>
<dd><p>scRNA-Seq data whose distribution is to be mapped onto <code>y</code>: either a
matrix of gene expression on log2+1 scale, or a 'cellMarkers' class object,
in which case the <code>$genemeans</code> list element is extracted.</p></dd>


<dt id="arg-y">y<a class="anchor" aria-label="anchor" href="#arg-y"></a></dt>
<dd><p>Reference scRNA-Seq data: either a matrix of gene expression on
log2+1 scale, or a 'cellMarkers' class object, in which case the
<code>$genemeans</code> list element is extracted.</p></dd>


<dt id="arg-n">n<a class="anchor" aria-label="anchor" href="#arg-n"></a></dt>
<dd><p>Number of quantiles to split <code>x</code> and <code>y</code>.</p></dd>


<dt id="arg-remove-noncoding">remove_noncoding<a class="anchor" aria-label="anchor" href="#arg-remove-noncoding"></a></dt>
<dd><p>Logical, whether to remove noncoding. This is a basic
filter which looks at the gene names (rownames) in both matrices and
removes genes containing "-" which are usually antisense or mitochondrial
genes, or "." which are either pseudogenes or ribosomal genes.</p></dd>


<dt id="arg-remove-zeros">remove_zeros<a class="anchor" aria-label="anchor" href="#arg-remove-zeros"></a></dt>
<dd><p>Logical, whether to remove zeros from both datasets.
This shifts the quantile relationships.</p></dd>


<dt id="arg-smooth">smooth<a class="anchor" aria-label="anchor" href="#arg-smooth"></a></dt>
<dd><p>Either "loess" or "lowess" which apply <code><a href="https://rdrr.io/r/stats/loess.html" class="external-link">loess()</a></code> or <code><a href="https://rdrr.io/r/stats/lowess.html" class="external-link">lowess()</a></code>
to smooth the QQ fitted line, or "ns" which uses natural splines via
<code>ns()</code>. With any other value no smoothing is applied. With no smoothing or
"loess/lowess", interpolation is limited to the original range of <code>x</code>, i.e.
it will clip for values &gt; <code>max(x)</code>.</p></dd>


<dt id="arg-span">span<a class="anchor" aria-label="anchor" href="#arg-span"></a></dt>
<dd><p>controls the degree of smoothing in <code><a href="https://rdrr.io/r/stats/loess.html" class="external-link">loess()</a></code> and <code><a href="https://rdrr.io/r/stats/lowess.html" class="external-link">lowess()</a></code>.</p></dd>


<dt id="arg-knots">knots<a class="anchor" aria-label="anchor" href="#arg-knots"></a></dt>
<dd><p>Vector of quantile points for knots for fitting natural splines.</p></dd>


<dt id="arg-respace">respace<a class="anchor" aria-label="anchor" href="#arg-respace"></a></dt>
<dd><p>Logical whether to respace quantile points so their x axis
density is more even. Can help spline fitting.</p></dd>


<dt id="arg-silent">silent<a class="anchor" aria-label="anchor" href="#arg-silent"></a></dt>
<dd><p>Logical whether to suppress messages.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A list object of class 'qqmap' containing:</p>
<dl><dt>quantiles</dt>
<dd><p>Dataframe containing matching quantiles of <code>x</code> and <code>y</code></p></dd>

<dt>map</dt>
<dd><p>A function of form <code>FUN(x)</code> where <code>x</code> can be supplied as a numeric
vector or matrix and the same type is returned. The function converts given
data points to the distribution of <code>y</code>.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>The conversion uses the function <code><a href="https://rdrr.io/r/stats/approxfun.html" class="external-link">approxfun()</a></code> which uses interpolation. It
is not designed to perform stepwise (exact) quantile transformation of every
individual datapoint.</p>
    </div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p><code><a href="https://rdrr.io/r/stats/approxfun.html" class="external-link">approxfun()</a></code></p></div>
    </div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by <a href="https://github.com/Zaoqu-Liu" class="external-link">Zaoqu Liu</a> and <a href="https://github.com/myles-lewis" class="external-link">Myles Lewis</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

    </footer></div>





  </body></html>

