<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>Simulate pseudo-bulk RNA-Seq — simulate_bulk • cellGeometry</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Simulate pseudo-bulk RNA-Seq — simulate_bulk"><meta name="description" content="Simulates pseudo-bulk RNA-Seq dataset using two modes. The first mode uses a
'cellMarkers' class object and a matrix of counts for the numbers of cells of
each cell subclass. This method converts the log2 gene means back for
each cell subclass back to count scale and then calculates pseudo-bulk count
values based on the cell amounts specified in samples. In the 2nd mode, a
single-cell RNA-Seq dataset is required, such as a matrix used as input to
cellMarkers(). Cells from the relevant subclass are sampled from the
single-cell matrix in the appropriate amounts based on samples, except that
sampling is scaled up by the factor times."><meta property="og:description" content="Simulates pseudo-bulk RNA-Seq dataset using two modes. The first mode uses a
'cellMarkers' class object and a matrix of counts for the numbers of cells of
each cell subclass. This method converts the log2 gene means back for
each cell subclass back to count scale and then calculates pseudo-bulk count
values based on the cell amounts specified in samples. In the 2nd mode, a
single-cell RNA-Seq dataset is required, such as a matrix used as input to
cellMarkers(). Cells from the relevant subclass are sampled from the
single-cell matrix in the appropriate amounts based on samples, except that
sampling is scaled up by the factor times."></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">cellGeometry</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.5.8</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../articles/intro.html">Get Started</a></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles"><li><a class="dropdown-item" href="../articles/algorithm.html">Mathematical Foundations</a></li>
    <li><a class="dropdown-item" href="../articles/visualization.html">Visualization Guide</a></li>
    <li><hr class="dropdown-divider"></li>
    <li><a class="dropdown-item" href="../articles/index.html">All articles</a></li>
  </ul></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-tutorials" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Tutorials</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-tutorials"><li><a class="dropdown-item" href="../articles/intro.html">Quick Start Guide</a></li>
    <li><a class="dropdown-item" href="../articles/real_world.html">Real-World Application</a></li>
    <li><a class="dropdown-item" href="../articles/brain_atlas.html">Large-Scale: Brain Atlas</a></li>
  </ul></li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json"></form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/Zaoqu-Liu/cellGeometry"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Simulate pseudo-bulk RNA-Seq</h1>
      <small class="dont-index">Source: <a href="https://github.com/Zaoqu-Liu/cellGeometry/blob/HEAD/R/simulate_bulk.R" class="external-link"><code>R/simulate_bulk.R</code></a></small>
      <div class="d-none name"><code>simulate_bulk.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>Simulates pseudo-bulk RNA-Seq dataset using two modes. The first mode uses a
'cellMarkers' class object and a matrix of counts for the numbers of cells of
each cell subclass. This method converts the log2 gene means back for
each cell subclass back to count scale and then calculates pseudo-bulk count
values based on the cell amounts specified in <code>samples</code>. In the 2nd mode, a
single-cell RNA-Seq dataset is required, such as a matrix used as input to
<code><a href="cellMarkers.html">cellMarkers()</a></code>. Cells from the relevant subclass are sampled from the
single-cell matrix in the appropriate amounts based on <code>samples</code>, except that
sampling is scaled up by the factor <code>times</code>.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">simulate_bulk</span><span class="op">(</span></span>
<span>  <span class="va">object</span>,</span>
<span>  <span class="va">samples</span>,</span>
<span>  <span class="va">subclass</span>,</span>
<span>  times <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  method <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"dirichlet"</span>, <span class="st">"unif"</span><span class="op">)</span>,</span>
<span>  alpha <span class="op">=</span> <span class="fl">1</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-object">object<a class="anchor" aria-label="anchor" href="#arg-object"></a></dt>
<dd><p>Either a 'cellMarkers' class object, or a single cell count
matrix with genes in rows and cells in columns, with rownames representing
gene IDs/symbols. The matrix can be a sparse matrix or DelayedMatrix.</p></dd>


<dt id="arg-samples">samples<a class="anchor" aria-label="anchor" href="#arg-samples"></a></dt>
<dd><p>An integer matrix of cell counts with samples in rows and
columns for each cell subclass in <code>object</code>. This can be generated using
<code><a href="generate_samples.html">generate_samples()</a></code>.</p></dd>


<dt id="arg-subclass">subclass<a class="anchor" aria-label="anchor" href="#arg-subclass"></a></dt>
<dd><p>Vector of cell subclasses matching the columns in <code>object</code>.
Only used if <code>object</code> is a single cell count matrix.</p></dd>


<dt id="arg-times">times<a class="anchor" aria-label="anchor" href="#arg-times"></a></dt>
<dd><p>Scaling factor to increase sampling of cells. Cell counts in
<code>samples</code> are scaled up by being multiplied by this number. Only used if
<code>object</code> is a single cell count matrix.</p></dd>


<dt id="arg-method">method<a class="anchor" aria-label="anchor" href="#arg-method"></a></dt>
<dd><p>Either "dirichlet" or "unif" to specify whether cells are
sampled based on the Dirichlet distribution with K = number of cells in
each subclass, or sampled uniformly. When cells are oversampled uniformly,
in the limit the summed gene expression tends to the arithmetic mean of the
subclass x sample frequency. Dirichlet sampling provides proper randomness
with sampling.</p></dd>


<dt id="arg-alpha">alpha<a class="anchor" aria-label="anchor" href="#arg-alpha"></a></dt>
<dd><p>Shape parameter for Dirichlet sampling.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>An integer count matrix with genes in rows and cell subclasses in
columns. This can be used as <code>test</code> with the <code><a href="deconvolute.html">deconvolute()</a></code> function.</p>
    </div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>The first method can give perfect deconvolution if the following settings are
used with <code><a href="deconvolute.html">deconvolute()</a></code>: <code>count_space = TRUE</code>, <code>convert_bulk = FALSE</code>,
<code>use_filter = FALSE</code> and <code>comp_amount = 1</code>.</p>
    </div>
    <div class="section level2">
    <h2 id="see-also">See also<a class="anchor" aria-label="anchor" href="#see-also"></a></h2>
    <div class="dont-index"><p><code><a href="generate_samples.html">generate_samples()</a></code> <code><a href="deconvolute.html">deconvolute()</a></code> <code><a href="add_noise.html">add_noise()</a></code></p></div>
    </div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by <a href="https://github.com/Zaoqu-Liu" class="external-link">Zaoqu Liu</a> and <a href="https://github.com/myles-lewis" class="external-link">Myles Lewis</a>.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.2.0.</p>
</div>

    </footer></div>





  </body></html>

